src/PluginHandler.cpp

```diff
@@ -0,0 +1,1065 @@
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,  USA.         *
+ ***************************************************************************
+ */
+#include <algorithm>
+#include <cstdio>
+#include <fstream>
+#include <iomanip>
+#include <memory>
+#include <ostream>
+#include <regex>
+#include <stdexcept>
+#include <streambuf>
+#include <sstream>
+#include <unordered_map>
+
+#include <wx/jsonreader.h>
+#include <wx/string.h>
+
+#include <archive.h>
+#include <archive_entry.h>
+typedef __LA_INT64_T la_int64_t;      //  "older" libarchive versions support
+
+#include <expat.h>
+
+#if defined(__MINGW32__) && defined(Yield)
+#undef Yield                 // from win.h, conflicts with mingw headers
+#endif
+
+#include "Downloader.h"
+#include "OCPNPlatform.h"
+#include "PluginHandler.h"
+#include "PluginPaths.h"
+#include "pluginmanager.h"
+#include "navutil.h"
+
+#ifdef XML_LARGE_SIZE
+#  if defined(XML_USE_MSC_EXTENSIONS) && _MSC_VER < 1400
+#    define XML_FMT_INT_MOD "I64"
+#  else
+#    define XML_FMT_INT_MOD "ll"
+#  endif
+#else
+#  define XML_FMT_INT_MOD "l"
+#endif
+
+#ifdef XML_UNICODE_WCHAR_T
+# include <wchar.h>
+# define XML_FMT_STR "ls"
+#else
+# define XML_FMT_STR "s"
+#endif
+
+#ifdef _WIN32
+static std::string SEP("\\");
+#else
+static std::string SEP("/");
+#endif
+
+#ifndef F_OK   // windows: missing unistd.h.
+#define F_OK 0
+#endif
+
+
+extern OCPNPlatform*  g_Platform;
+extern PlugInManager* g_pi_manager;
+extern wxString       g_winPluginDir;
+extern MyConfig*      pConfig;
+extern OCPNPlatform*  g_Platform;
+
+struct parse_ctx {
+    std::vector<OcpnPlugin> plugins;
+    std::unique_ptr<OcpnPlugin> plugin;
+    std::string buff;
+};
+
+
+static bool endswith(const std::string& str, const std::string& suffix)
+{
+    return str.size() >= suffix.size()
+        && 0 == str.compare(str.size()-suffix.size(), suffix.size(), suffix);
+}
+
+
+static bool startswith(const std::string& str, const std::string& prefix)
+{
+    return strncmp(str.c_str(), prefix.c_str(), prefix.size()) == 0;
```

Comment:
Should this be: ```prefix.size() <= str.size() && strncmp(...);```?

Reply:
Perhaps even ```prefix.size() <= str.size() && std::equal(prefix.begin(), prefix.end(), str.begin(), str.begin()+prefix.size());``` to avoid c_str?

Reply:
First one fixed.  The last ... yes, aayre are right. Will fix.

Reply:
Fixed in  ocpn_utils.cpp, at last.